src/app/qgisapp.cpp

```diff
@@ -5853,6 +5855,117 @@ void QgisApp::saveMapAsImage()
 
 } // saveMapAsImage
 
+void QgisApp::saveMapAsPdf()
+{
+  QList< QgsMapDecoration * > decorations;
+  QString activeDecorations;
+  Q_FOREACH ( QgsDecorationItem *decoration, mDecorationItems )
+  {
+    if ( decoration->enabled() )
+    {
+      decorations << decoration;
+      if ( activeDecorations.isEmpty() )
+        activeDecorations = decoration->name().toLower();
+      else
+        activeDecorations += QString( ", %1" ).arg( decoration->name().toLower() );
+    }
+  }
+
+  QgsMapSaveDialog dlg( this, mMapCanvas, activeDecorations, QgsMapSaveDialog::Pdf );
+  if ( !dlg.exec() )
+    return;
+
+  QgsSettings settings;
+  QString lastUsedDir = settings.value( QStringLiteral( "UI/lastSaveAsImageDir" ), QDir::homePath() ).toString();
+  QString myFileName = QFileDialog::getSaveFileName( this, tr( "Save map as" ), lastUsedDir, tr( "PDF Format" ) + " (*.pdf *.PDF)" );
+  if ( myFileName != QLatin1String( "" ) )
+  {
+    QgsMapSettings ms = QgsMapSettings();
+    ms.setDestinationCrs( QgsProject::instance()->crs() );
+    ms.setExtent( dlg.extent() );
+    ms.setOutputSize( dlg.size() );
+    ms.setOutputDpi( dlg.dpi() );
+    ms.setBackgroundColor( mMapCanvas->canvasColor() );
+    ms.setRotation( mMapCanvas->rotation() );
+    ms.setLayers( mMapCanvas->layers() );
+
+    QPrinter *printer = new QPrinter();
+    printer->setOutputFileName( myFileName );
+    printer->setOutputFormat( QPrinter::PdfFormat );
+    printer->setOrientation( QPrinter::Portrait );
+    printer->setPaperSize( dlg.size(), QPrinter::DevicePixel );
+    printer->setPageMargins( 0, 0, 0, 0, QPrinter::DevicePixel );
+
+    QPainter *p = new QPainter();
+    QImage *image = nullptr;
+    if ( dlg.saveAsRaster() )
+    {
+      image = new QImage( dlg.size(), QImage::Format_ARGB32 );
+      p->begin( image );
+    }
+    else
+    {
+      p->begin( printer );
+    }
+
+    QgsMapRendererTask *mapRendererTask = new QgsMapRendererTask( ms, p );
+
+    if ( dlg.drawAnnotations() )
+    {
+      mapRendererTask->addAnnotations( QgsProject::instance()->annotationManager()->annotations() );
+    }
+
+    if ( dlg.drawDecorations() )
+    {
+      mapRendererTask->addDecorations( decorations );
+    }
+
+    mapRendererTask->setSaveWorldFile( dlg.saveWorldFile() );
+
+    connect( mapRendererTask, &QgsMapRendererTask::renderingComplete, this, [ = ]
```

Comment:
I'd suggest explicitly capturing image here ([this, image]). It makes it a bit clearer how this functions